### ISSUE TITLE
comment for finalization added to documentation for derived-type

### ISSUE THREAD

**Issue by jdonners** [cmacmackin/ford#109]
Hi,

sorry to bother you again. I hope I'm more awake this time.
It looks like the comment for a finalization procedure of a derived-type
is added to the documentation for the derived-type. E.g. see:

http://fortran-foss-programmers.github.io/FOODIE/type/adams_bashforth_integrator.html

It looks like a minor issue, but I don't understand the code well enough to fix it just yet.

**Comment by szaghi** [comment]
@jdonners do not worry, I am replaying before Chris just because my time zone wakes me before Chris :smile: 

Indeed, I do not understand *where* is this issue: yes, the final TBP is documented, is this the issue? I think that happens because I like to document all, thus I use `display: public protected private` in my main FORD file.

**Comment by jdonners** [comment]
I agree that it's good to document all (even though I haven't done that yet myself). The issue is that the source code in `foodie_integrator_adams_bashforth.f90`looks like:
```fortran
type :: adams_bashforth_integrator
...
    final                         :: finalize        !< Finalize object.
```

where the comment 'Finalize object.' is inserted at the end of the description for the derived-type (Search for 'steps; Finalize object.' in the page mentioned above), not in the description of the finalization procedures.

**Comment by cmacmackin** [comment]
Yes, it would appear that I forgot to have FORD scan for documentation 
after a `final` statement. Rather an odd little bug.

On 17/02/16 09:54, John Donners wrote:
>
> I agree that it's good to document all (even though I haven't done 
> that yet myself). The issue is that the source code in 
> |foodie_integrator_adams_bashforth.f90|looks like:
>
> type::  adams_bashforth_integrator
> ...
>      final::  finalize!< Finalize object.
>
> where the comment 'Finalize object.' is inserted at the end of the 
> description for the derived-type (Search for 'steps; Finalize object.' 
> in the page mentioned above), not in the description of the 
> finalization procedures.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cmacmackin/ford/issues/109#issuecomment-185126821>.
>

-- 
Chris MacMackin
cmacmackin.github.io <http://cmacmackin.github.io>

**Comment by szaghi** [comment]
@jdonners Ok, I understand now. 

Maybe, this FOODIE example is misleading because I have used the same docstring for two different *kind* of documentation. 

I could be wrong, but I guessed that

### docstring at the type-contains level
```fortran
type :: adams_bashforth_integrator
...
    final                         :: finalize        !< Finalize object.
```
This docstring should go into the *type* doc page, e.g. [this](http://fortran-foss-programmers.github.io/FOODIE/type/adams_bashforth_integrator.html)

### docstring at the TPB level
```fortran
elemental subroutine finalize(self)
  !< Finalize object.
...
```
This doc string should go into the TPB doc page, e.g. [this](http://fortran-foss-programmers.github.io/FOODIE/proc/finalize.html)

**Comment by szaghi** [comment]
@cmacmackin @jdonners Sorry, after the comment of Chris I am now confused...

**Comment by cmacmackin** [comment]
For some reason, FORD is seeing the

```
type :: adams_bashforth_integrator
...
    final                         :: finalize        !< Finalize object.
```
comment and thinking that "Finalize object." is part of the documentation of the derived type. As such, it appends it to the end of that set of documentation and doesn't show it with the binding of the finalization procedure. Confusingly, it does show them within the documentation of the subroutine _performing_ the finalization, because there is the same documentation there.

**Comment by cmacmackin** [comment]
Okay, I've fixed this now. Incidentally, @szaghi, did you realize that there is no implementation of the procedure `bctor_r16p` in IR_precision.f90 (see [this line](https://github.com/szaghi/IR_Precision/blob/master/src/IR_Precision.f90#L234))? I was running FORD on FOODIE and it was having trouble with the IR_precision module, so I did a bit of digging to find out why. There was a bug in FORD which I fixed, but I also discovered this error. Incidentally, since IR_precision uses the preprocessor you should really give it the extension `.F90`.

**Comment by szaghi** [comment]
@cmacmackin thank you for pointing out this bug on IR_precision. I do not use r16p, it was add for a friend that worked on relativistic math, but they are years that no-one use r16p... maybe it is the time to sanitize the module...

Thank you very much!

**Comment by cmacmackin** [comment]
Fixed in  [v4.5.2](https://github.com/cmacmackin/ford/releases/tag/v4.5.2). Can be installed with `pip install --upgrade ford`.
